File: DeflateCS/Deflate.cs
// public domain

using System;
using System.Collections.Generic;
using System.IO;

namespace DeflateCS
{
    public static class Deflate
    {
        public const int maxlen = 258;
        public const int maxdist = 32768;

        public static int[] litexlens = new int[286];
        public static int[] litlens = new int[286];
        public static int[] litindex = new int[maxlen - 2];
        public static int[] distexlens = new int[30];
        public static int[] distlens = new int[30];
        public static int[] distindex = new int[maxdist];
        public static byte[, ,] sl = new byte[8192, 8, 3];
        public static byte[] rev = new byte[256];

        static Deflate()
        {
            for (int i = 265; i <= 284; i++)
                litexlens[i] = (i - 261) >> 2;

            int v = 3;
            for (int i = 257; i <= 284; i++)
            {
                litlens[i] = v;
                int p2 = 1;
                for (int j = 1; j <= litexlens[i]; j++)
                    p2 <<= 1;
                for (int j = 1; j <= p2; j++, v++)
                    litindex[v - 3] = i;
            }
            litlens[285] = maxlen;
            litindex[maxlen - 3] = 285;

            for (int i = 4; i < 30; i++)
                distexlens[i] = (i - 2) >> 1;

            v = 1;
            for (int i = 0; i < 30; i++)
            {
                distlens[i] = v;
                int p2 = 1;
                for (int j = 1; j <= distexlens[i]; j++)
                    p2 <<= 1;
                for (int j = 1; j <= p2; j++, v++)
                    distindex[v - 1] = i;
            }

            for (int i = 0; i < 8192; i++)
            {
                int p2 = 1;
                for (int j = 0; j < 8; j++)
                {
                    v = i * p2;
                    sl[i, j, 0] = (byte)(v & 255);
                    sl[i, j, 1] = (byte)((v >> 8) & 255);
                    sl[i, j, 2] = (byte)(v >> 16);
                    p2 += p2;
                }
            }

            for (int i = 0; i < 256; i++)
            {
                int p2 = 1;
                int p2r = 128;
                int b = 0;
                for (int j = 0; j < 8; j++)
                {
                    if ((i & p2) != 0) b += p2r;
                    p2 <<= 1;
                    p2r >>= 1;
                }
                rev[i] = (byte)b;
            }
        }

        public static byte[] GetCompressBytes(Stream sin)
        {
            var ms = new MemoryStream();
            var w = new DeflateWriter();
            w.Compress(sin, ms);
            ms.Close();
            return ms.ToArray();
        }
    }
}

File: DeflateVB/Deflate.vb
' public domain

Imports System.IO

Module Deflate
    Public Const maxlen = 258, maxdist = 32768
    Public litexlens(285), litlens(285), litindex(maxlen - 3) As Integer
    Public distexlens(29), distlens(29), distindex(maxdist - 1) As Integer
    Public sl(8191, 7, 2) As Byte, rev(255) As Byte

    Public Sub Init()
        Dim I, J, V, P2 As Integer
        Dim P2R As Byte, B As Byte

        For I = 265 To 284
            litexlens(I) = (I - 261) >> 2
        Next I

        V = 3
        For I = 257 To 284
            litlens(I) = V
            P2 = 1
            For J = 1 To litexlens(I)
                P2 <<= 1
            Next J
            For J = 1 To P2
                litindex(V - 3) = I
                V += 1
            Next J
        Next I
        litlens(285) = maxlen
        litindex(maxlen - 3) = 285

        For I = 4 To 29
            distexlens(I) = (I - 2) \ 2
        Next I

        V = 1
        For I = 0 To 29
            distlens(I) = V
            P2 = 1
            For J = 1 To distexlens(I)
                P2 <<= 1
            Next
            For J = 1 To P2
                distindex(V - 1) = I
                V += 1
            Next J
        Next I

        For I = 0 To 8191
            P2 = 1
            For J = 0 To 7
                V = I * P2
                sl(I, J, 0) = V And 255
                sl(I, J, 1) = (V >> 8) And 255
                sl(I, J, 2) = V >> 16
                P2 = P2 + P2
            Next J
        Next I

        For I = 0 To 255
            P2 = 1
            P2R = 128
            B = 0
            For J = 0 To 7
                If I And P2 Then B = B + P2R
                P2 <<= 1
                P2R >>= 1
            Next J
            rev(I) = B
        Next I
    End Sub

    Public Sub Compress(sin As Stream, sout As Stream)
        Dim w = New DeflateWriter
        w.Compress(sin, sout)
    End Sub

    Public Function GetCompressBytes(sin As Stream) As Byte()
        If rev(255) = 0 Then Init()
        Dim ms = New MemoryStream
        Compress(sin, ms)
        ms.Close()
        GetCompressBytes = ms.ToArray()
    End Function
End Module
